### PR TITLE
Add FNO4D (4D xFNO) operator to experimental xdeeponet

### LINKED ISSUE #1708: 🚀[FEA]: Add 4D FNO operator (FNO4D) to experimental xdeeponet
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem you would like to solve.

The merged xDeepONet core (`physicsnemo.experimental.models.xdeeponet.DeepONet`)
is dimension-capped at `dimension=2|3` and cannot express a **4D Fourier
Neural Operator** over `(B, X, Y, Z, T, C)` (3D space + time), which is needed
for volumetric time-dependent problems (e.g. reservoir simulation /
Norne-style datasets).

Proposal: add `FNO4D` / `FNO4DWrapper` alongside `DeepONet` in the same
`xdeeponet` package, as additive `physicsnemo.Module` classes (without
modifying the merged core), reusing the existing `SpectralConv4d` /
`ConvNdFCLayer` / `ConvNdKernel1Layer` primitives. The wrapper provides
automatic spectral padding and optional autoregressive time-axis extension.
Output should be bit-identical to the Neural Operator Factory `FNO4D` source.

### Describe any alternatives you have considered

- The 3D FNO / Conv-FNO / U-FNO operators are **not** part of this request:
  they are already expressible as `DeepONet(trunk=None, dimension=3)` with a
  Fourier/UNet/Conv `SpatialBranch`, so adding them as separate classes would
  be redundant.
- Extending the merged `DeepONet` core to `dimension=4`: rejected to avoid
  modifying already-shipped code and because the 4D variant has distinct
  numerics (e.g. no skip-branches; bit-parity with the NOF source).

### PR DESCRIPTION
**Pull request by wdyab**
# PhysicsNeMo Pull Request

## Description

closes #1708

Adds the **4D Fourier Neural Operator** (`FNO4D`) — 3D space + time —
alongside `DeepONet` in `physicsnemo.experimental.models.xdeeponet`, as
additive, standalone `physicsnemo.Module` classes. The merged xDeepONet core
(`deeponet.py` / `branches.py`) is **not modified**.

New classes:

- `FNO4D` + `FNO4DWrapper` — pure 4D FNO over `(B, X, Y, Z, T, C)`. This is
  the higher-dimensional operator the dimension-capped (`dimension=2|3`)
  `DeepONet` core cannot express. It uses the dimension-agnostic
  `SpectralConv4d` / `ConvNdFCLayer` / `ConvNdKernel1Layer` primitives (no
  `nn.Conv4d` exists, so no U-Net/Conv skip-branch in 4D). The wrapper adds
  automatic right-side spectral padding and optional autoregressive time-axis
  extension via `target_times`.

A single additive helper (`compute_right_pad_to_multiple_per_dim`) is appended
to the existing `_padding.py`; the new classes are exported from the package
`__init__`.

**3D FNO / Conv-FNO / U-FNO are intentionally not added as separate classes.**
They are redundant with `DeepONet(trunk=None, dimension=3)` plus a
`SpatialBranch` composed of Fourier / UNet / Conv layers over the
`(H, W, T)` axes, which already reproduces them. Only the genuinely-new 4D
operator is added here.

This is a behavior-preserving port of the Neural Operator Factory `FNO4D`
source (`examples/reservoir_simulation/neural_operator_factory/models/xfno.py`);
forward output is **bit-identical** to the original (`rtol=0, atol=0`).

Tests mirror the merged xDeepONet suite format: `test_xfno.py` (constructor +
public attributes, forward non-regression vs committed golden `.pth`,
checkpoint round-trip, gradient flow, `torch.compile`, and time-axis
extension), driven by a `_FIXTURE_REGISTRY` with a `_generate_xfno_goldens.py`
regeneration script and two committed golden fixtures.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies

None. All required primitives (`SpectralConv4d`, `ConvNdFCLayer`,
`ConvNdKernel1Layer`) already ship in `physicsnemo.nn`.

### CODE REVIEW

**Comment by wdyab** [comment]
@melo-gonzo for your attention and review

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+class _FNO4DMetaData(ModelMetaData):
+    """PhysicsNeMo model metadata for :class:`FNO4D`."""
+
+
+@dataclass
+class _FNO4DWrapperMetaData(ModelMetaData):
+    """PhysicsNeMo model metadata for :class:`FNO4DWrapper`."""
+
+
+class FNO4D(Module):
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: Most of this entire class reimplements the [`FNO4DEncoder`](https://github.com/NVIDIA/physicsnemo/blob/980bf9d8bcaab3c352337f77de1b7363ed2694c9/physicsnemo/nn/module/fno_layers.py#L672). It would be worth consolidating this duplicate code and only including what is strictly new. The 4d FNO can be instantiated by `physicsnemo.models.fno.FNO(dimension=4)`

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+        layers_list: list[nn.Module] = []
+        in_ch = width
+        for _ in range(num_layers):
+            layers_list.append(ConvNdFCLayer(in_ch, hidden_width))
+            layers_list.append(self.activation_fn)
+            in_ch = hidden_width
+        layers_list.append(ConvNdFCLayer(hidden_width, out_channels))
+        return nn.Sequential(*layers_list)
+
+    def _create_meshgrid(self, shape: list, device: torch.device) -> Tensor:
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: See [existing](https://github.com/NVIDIA/physicsnemo/blob/980bf9d8bcaab3c352337f77de1b7363ed2694c9/physicsnemo/nn/module/fno_layers.py#L836) implementation

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+            )
+            self.conv_1x1s.append(ConvNdKernel1Layer(self.width, self.width))
+
+        self.decoder = self._build_decoder_network(
+            width, out_channels, decoder_layers, decoder_width
+        )
+
+    # ---------------------------------------------------------------- build helpers
+
+    def _build_lifting_network(
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: See [here](https://github.com/NVIDIA/physicsnemo/blob/980bf9d8bcaab3c352337f77de1b7363ed2694c9/physicsnemo/nn/module/fno_layers.py#L759) in `physicsnemo.nn.module.fno_layers.py`.

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+            lift_in_channels, width, lifting_layers
+        )
+
+        self.spectral_convs = nn.ModuleList()
+        self.conv_1x1s = nn.ModuleList()
+        for _ in range(num_fno_layers):
+            self.spectral_convs.append(
+                SpectralConv4d(self.width, self.width, modes1, modes2, modes3, modes4)
+            )
+            self.conv_1x1s.append(ConvNdKernel1Layer(self.width, self.width))
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: Much of this is covered [here](https://github.com/NVIDIA/physicsnemo/blob/980bf9d8bcaab3c352337f77de1b7363ed2694c9/physicsnemo/nn/module/fno_layers.py#L771)

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+        for layer_idx in range(self.num_fno_layers):
+            x1 = self.spectral_convs[layer_idx](x)
+            x2 = self.conv_1x1s[layer_idx](x)
+            if layer_idx < self.num_fno_layers - 1:
+                x = self.activation_fn(x1 + x2)
+            else:
+                x = x1 + x2
+
+        # Decoder.
+        x = self.decoder(x)  # (B, C_out, X, Y, Z, T)
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: Forward in 4dFNO [here](https://github.com/NVIDIA/physicsnemo/blob/980bf9d8bcaab3c352337f77de1b7363ed2694c9/physicsnemo/nn/module/fno_layers.py#L797)

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+        hidden_width = width // 2
+        layers_list.append(ConvNdFCLayer(in_channels, hidden_width))
+        layers_list.append(self.activation_fn)
+        for _ in range(num_layers - 2):
+            layers_list.append(ConvNdFCLayer(hidden_width, hidden_width))
+            layers_list.append(self.activation_fn)
+        layers_list.append(ConvNdFCLayer(hidden_width, width))
+        return nn.Sequential(*layers_list)
+
+    def _build_decoder_network(
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: Decoder net building in [`physicsnemo.models.fno](https://github.com/NVIDIA/physicsnemo/blob/980bf9d8bcaab3c352337f77de1b7363ed2694c9/physicsnemo/models/fno/fno.py#L173)

**Review by melo-gonzo — COMMENTED** [review]
Hey @wdyab, thank you for this PR, it is well documented and with quality code. It does however look like much of the `fno4d.py` file is duplicate to what PhysicsNeMo already offers between `physicsnemo.models.fno` and `physicsnemo.nn.modules.fno_layers`. Please review this main class, (`FNO4D`) and the comments I have left. If you believe something is notably different between what currently exists and what you've proposed, please do acknowledge those differences! 

I may suggest keeping the `FNO4DWrapper` and instantiating the existing 4dFNO in the `__init__`. 

Otherwise, the additional utilities and additions look good to me.

**Comment by wdyab** [comment]
Thanks @melo-gonzo, great catch - you're right, the FNO4D core was duplicating FNO4DEncoder / physicsnemo.models.fno.FNO, I gues I was stuck thinking about my NOF code that I didn't realize that this is no longer needed. I've removed it and FNO4DWrapper now instantiates physicsnemo.models.fno.FNO(dimension=4) in its __init__, as you suggested.

The only thing kept is what the native FNO doesn't offer: the autoregressive time-axis extension (target_times → predict a K-step horizon), plus the channel-last (B, X, Y, Z, T, C) convention and trailing-channel squeeze. Domain padding is now delegated to the native FNO, and the 4D core is covered by the existing test/models/fno suite.

Pushed in a436a12b. Thanks for the thorough review!

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 7]
```diff
@@ -14,48 +14,26 @@
...
+  explicit forecast horizon via ``target_times``, the time axis is
+  right-replicate-padded so the operator predicts ``K`` future steps, and the
+  output is cropped back to those ``K`` steps.  It also adopts the channel-last
+  :math:`(B, X, Y, Z, T, C)` input/output convention and squeezes a trailing
+  unit channel.
+
+This consolidation follows reviewer feedback on the PR: the standalone 4D FNO
+core duplicated ``physicsnemo.models.fno`` / ``physicsnemo.nn.module.fno_layers``
+and was removed.  The autoregressive time-axis extension (originally from the
+Neural Operator Factory ``FNO4DNet`` wrapper) is the only piece retained here.
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: You can remove these types of comments directly from the code. The PR history will show our discourse and we dont need to explicitly track these changes in the python file comments.

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 8]
```diff
@@ -14,48 +14,26 @@
...
-on the Neural Operator Factory (NOF) feature branch.  Every option is
-preserved verbatim — same name, same default, same semantics.  Output is
-**bit-identical** to the NOF source under matching seed / config / input.
+"""4D xFNO wrapper (3D space + time) for the xDeepONet family.
+
+The 4D Fourier Neural Operator itself is **not** re-implemented here: it is
+provided by the library model :class:`physicsnemo.models.fno.FNO` with
+``dimension=4`` (backed by
+:class:`physicsnemo.nn.module.fno_layers.FNO4DEncoder`).  This module only
+adds the one capability that the core ``FNO`` does not provide:
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: Similar note about comments here. Not necessary to explicity state what we are *not* reimplementing. It would suffice to say we are using `physicsnemo.models.fno.FNO` and `physicsnemo.nn.module.fno_layers.FNO4DEncoder`.

File: physicsnemo/experimental/models/xdeeponet/fno4d.py [hunk 9]
```diff
@@ -513,55 +153,42 @@ class FNO4DWrapper(Module):
...
-                )
-            self.padding = (pad_list + [0] * (4 - len(pad_list)))[:4]
+            self.time_modes = int(num_fno_modes)
 
-        # Cached so the time-extension branch in forward can size the
-        # right-pad without re-inspecting the inner model.
-        self.time_modes = modes4
-
-        self.fno4d = FNO4D(
+        # The 4D FNO core is the library model -- not re-implemented here.
```
Comment (melo-gonzo) [inline physicsnemo/experimental/models/xdeeponet/fno4d.py]: Similar to other comments, no need to explicitly state what is not re-implemented.

**Review by melo-gonzo — APPROVED** [review]
Thanks for the updates @wdyab. I added a few more minor notes about the content in comments - please do a sweep over the comments and docstrongs. After this it should be good to merge!

**Comment by wdyab** [comment]
Thanks @melo-gonzo! Swept the docstrings/comments in fno4d.py. 
Pushed in a6788d38. Thanks for the review!

**Comment by melo-gonzo** [comment]
/ok to test [982a649](https://github.com/NVIDIA/physicsnemo/pull/1707/commits/982a6494dd58d7fc762cfbe95af46b18a9d052a3)
